### PR TITLE
ADD GLU activation function

### PR DESCRIPTION
**Pull request by bn-c**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Saw an activation function I want to use, but haven't been implemented yet

### Changes

Added the Gated Linear Unit activation function from [`torch.nn.GLU`](https://docs.pytorch.org/docs/stable/generated/torch.nn.GLU.html#torch.nn.GLU): 
- GLU module. 
- glu activation functon.

### Testing

~~Required feedback on how to organize and implement tests~~
Added test with random data.

### CODE REVIEW

File: crates/burn-core/src/nn/glu.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        Self { dim }
+    }
+
+    /// Applies the gated linear unit function.
+    ///
+    /// GLU(a,b)=a⊗σ(b) where `a` is the first half of the input matrices and `b` is the second half.
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The input tensor. With shape [∗1,N,∗2] where * means, any number of additional dimensions
```
Comment (laggui) [inline crates/burn-core/src/nn/glu.rs]: I think we can skip the shape description, the only relevant parts are that
1) The input tensor size along `dim` must be divisible by 2.
2) The output tensor has the same shape as the input, except the size along `dim` is halved.

Feels more clear imo.

File: crates/burn-tensor/src/tensor/activation/base.rs [hunk 2]
```diff
@@ -259,3 +259,26 @@ pub fn mish<const D: usize, B: Backend>(tensor: Tensor<B, D>) -> Tensor<B, D> {
...
+/// $$
+/// where $a$ is the first half of the input matrices and $b$ is the second half.
+///
+/// # Arguments
+///
+/// * `tensor` - The input tensor. With shape [∗1,N,∗2] where * means, any number of additional dimensions
+/// * `dim` - The dimension on which to split the input.
+///
+/// # Returns
+/// Output tensor with shape [∗1​,M,∗2​] where M=N/2
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/activation/base.rs]: Same

**Review by laggui — CHANGES_REQUESTED** [review]
Thanks for adding the activation 🙂 

Implementation looks good! But it's missing tests.

See the tests in [`burn-tensor/src/tests/activation`](https://github.com/tracel-ai/burn/tree/main/crates/burn-tensor/src/tests/activation) for reference. Should add one for reference with `glu`. You can use pytorch as a reference to compute the expected output for a given input.

Some notes (not actionable, just for reference):
For other ops we might add autodiff tests, but again here this is a simple composition of existing ops. So not required.

Depending on the activation we might add it to the [`ActivationOps`](https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/ops/activation.rs), but in this case it's a simple activation composed of other ops. So I don't think it's necessary (just like other activations).

File: crates/burn-tensor/src/tensor/activation/base.rs (hunk 2 above)
Comment (bn-c) [reply crates/burn-tensor/src/tensor/activation/base.rs]: fixed

File: crates/burn-core/src/nn/glu.rs (hunk 1 above)
Comment (bn-c) [reply crates/burn-core/src/nn/glu.rs]: fixed

File: crates/burn-core/src/nn/glu.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    ///
+    /// **Note**:
+    /// * The input tensor along `dim` must have an even size along the specified dimension. N is divisible by 2.
+    /// * Negative indices for `dim` are not supported (unlike PyTorch's nn.GLU).
+    ///
+    /// ### Arguments
+    /// * `tensor` - The input tensor. With shape `[∗1,N,∗2]` where `*` means, any number of additional dimensions
+    ///
+    /// ### Returns
+    /// * Output tensor with shape `[∗1​,M,∗2​]` where M=N/2
```
Comment (laggui) [inline crates/burn-core/src/nn/glu.rs]: This is what I meant by skipping shape description, which was a little non-standard and ambiguous.

```suggestion
    /// **Note**:
    /// * The size of the input tensor along `dim` must be divisible by 2.
    ///
    /// ### Arguments
    /// * `tensor` - The input tensor.
    ///
    /// ### Returns
    /// * A tensor with the same shape as the input, except the size along `dim` is halved.
```

I also removed the note on negative indexing as I feel this is pretty clear with the use of `dim: usize`.

File: crates/burn-tensor/src/tensor/activation/base.rs [hunk 4]
```diff
@@ -259,3 +259,32 @@ pub fn mish<const D: usize, B: Backend>(tensor: Tensor<B, D>) -> Tensor<B, D> {
...
+///
+/// **Note**:
+/// * The input tensor along `dim` must have an even size along the specified dimension. N is divisible by 2.
+/// * Negative indices for `dim` are not supported (unlike PyTorch's nn.GLU).
+///
+/// ### Arguments
+/// * `tensor` - The input tensor. With shape `[∗1,N,∗2]` where `*` means, any number of additional dimensions
+///
+/// ### Returns
+/// * Output tensor with shape `[∗1​,M,∗2​]` where M=N/2
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/activation/base.rs]: Reflected changes from above

```suggestion
/// **Note**:
/// * The size of the input tensor along `dim` must be divisible by 2.
///
/// ### Arguments
/// * `tensor` - The input tensor.
///
/// ### Returns
/// * A tensor with the same shape as the input, except the size along `dim` is halved.
```

**Review by laggui — APPROVED** [review]
Thanks for adding a numerical test 🙂 

I think the shape documentation can be clearer if we simply use natural language to describe that.

Approving in advance
